### PR TITLE
PairK

### PR DESCRIPTION
**Pull request by ferranpujolcamins**
## Related issues

First step for #643 

## Goal

Add PairK type, product of functors.

### CODE REVIEW

File: Sources/Bow/Data/PairK.swift [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+
+// MARK: Instance of HashableK for PairK.
+extension PairKPartial: HashableK where F: HashableK, G: HashableK {
+    public static func hash<A>(_ fa: PairKOf<F, G, A>, into hasher: inout Hasher) where A : Hashable {
+        hasher.combine(fa^.run)
+    }
+}
+
+// MARK: Instance of Invariant for PairK.
+extension PairKPartial: Invariant where F: Functor, G: Functor {}
```
Comment (truizlop) [inline Sources/Bow/Data/PairK.swift:85]: This could be relaxed to F and G being Invariant.

File: Sources/Bow/Data/PairK.swift [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+    public static func ap<A, B>(_ ff: PairKOf<F, G, (A) -> B>, _ fa: PairKOf<F, G, A>) -> PairKOf<F, G, B> {
+        PairK(
+            F.ap(ff^.first, fa^.first),
+            G.ap(ff^.second, fa^.second)
+        )
+    }
+}
+
+// MARK: Instance of Selective for PairK
+extension PairKPartial: Selective where F: Monad, G: Monad {}
```
Comment (truizlop) [inline Sources/Bow/Data/PairK.swift:109]: This could be relaxed to F and G being Selective.

File: Sources/Bow/Data/PairK.swift [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+    public static func tailRecM<A, B>(_ a: A, _ f: @escaping (A) -> PairKOf<F, G, Either<A, B>>) -> PairKOf<F, G, B> {
+        PairK(
+            F.tailRecM(a, { f($0)^.first }),
+            G.tailRecM(a, { f($0)^.second })
+        )
+    }
+}
+
+// MARK: Instance of FunctorFilter for PairK
+extension PairKPartial: FunctorFilter where F: MonadFilter, G: MonadFilter {}
```
Comment (truizlop) [inline Sources/Bow/Data/PairK.swift:129]: This could be relaxed to F and G being FunctorFilter.

**Review by truizlop — CHANGES_REQUESTED** [review]

**Comment by ferranpujolcamins** [comment]
Fixed

**Review by truizlop — APPROVED** [review]
Approved, pending fixing merge conflicts.

**Comment by ferranpujolcamins** [comment]
Conflicts fixed

**Comment by ferranpujolcamins** [comment]
🚀

**Pull request merged**